.github/actions/make_init/action.yml

```diff
@@ -0,0 +1,77 @@
+name: "Initialize Virtual Env"
+runs:
+  using: composite
+  steps:
+    - name: Restore nvm, pre-commit cache, & node modules from cache
+      id: cache-nvm-node-modules
+      uses: actions/cache@v3
+      with:
+        path: |
+          ~/.nvm
+          ~/.cache
+          frontend/node_modules
+        key: v13-nvm-node-modules-${{ hashFiles('**/yarn.lock.md5') }}-${{ hashFiles('**/.pre-commit-config.yaml') }}
+    - name: Install nvm, pre-commit cache and node_modules
+      if: steps.cache-nvm-node-modules.cache-hit != 'true'
+      run: |
+        if [ ! -d ~/.nvm ] ; then
+          cd ..
+          export NVM_DIR="$HOME/.nvm"
+          [ -s "$NVM_DIR/nvm.sh" ] && \. "$NVM_DIR/nvm.sh"  # This loads nvm
+          mkdir $NVM_DIR
+          curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.35.3/install.sh | bash
+          echo "source $NVM_DIR/nvm.sh --install" >> "$HOME/.bash_profile"
+          cd -
+          source $HOME/.bash_profile
+          echo "lts/*" > .nvmrc
+          nvm install
+        fi
+        npm install --location=global yarn
+        if [ ! -d frontend/node_modules ] ; then
+          source "$HOME/.nvm/nvm.sh"
+          make react-init
+        fi
+      shell: bash -ileo pipefail {0}
+    - name: Install pyodbc dependencies
+      run: |
+        sudo apt-get install -y unixodbc-dev
+      shell: bash -ileo pipefail {0}
+    - name: Install graphviz dependencies
+      run: |
+        sudo apt update
+        sudo apt-get install -y libgvc6
+      shell: bash -ileo pipefail {0}
+    - name: Restore virtualenv from cache
+      id: cache-virtualenv
+      uses: actions/cache@v3
+      with:
+        path: venv
+        key: v13-python-venv-${{ hashFiles('**/python_cache_key.md5') }}
+    - name: Create Virtual Env
+      if: steps.cache-virtualenv.outputs.cache-hit != 'true'
+      run: |
+        python -m venv venv
+        source venv/bin/activate
+        pip install --upgrade pip
+        make setup
+        make pipenv-install
+        deactivate
+
+        # Add 'activate venv' to $HOME/.bash_profile. This means that our venv will be active
+        # for the remainder of the job ($HOME/.bash_profile is evaluated at each step).
+        echo 'source venv/bin/activate' >> $HOME/.bash_profile
+      shell: bash -ileo pipefail {0}
+    - name: Generate Protobufs
+      run: |
+        source venv/bin/activate
```

Comment:
Is it needed?

Reply:
+1

Reply:
Agreed that this shouldn't be needed, but steps will fail without it. For example, removing this instance of sourcing the venv produces a Thus it is unfortunately repeated throughout the workflows.

Reply:
Can you check this solution?
https://github.com/streamlit/streamlit/pull/5158/files#r948395439

Reply:
Oh whoops, so the issue is probably that we're only adding the line to `.bash_profile` on a cache miss, so on a cache hit the venv isn't activated when a new shell session is initiated.

Reworking things so that it always gets added to `.bash_profile` should fix things